README.md
# kvs - Secure Key Value Store

Have you ever implemented an application that had to store data persistently?
Have you ever been afraid of the confidentiality of this information?

Using **kvs** you can store your data protected by AES-256-GCM using hardware bound passwords.

The key-value-store-daemon **kvsd** takes care of storing your data and encrypting it at rest.
It provides a simple to use [gRPC](https://gRPC.io) interface that can be used in nearly any programming language.
If you want to run **kvsd** on a remote server you can enable TLS protection to ensure confidentialty during data transfer.

To be able to use the key-value store in shell scripts, **kvsc** provides a commandline interface for accessing the key-value-store.

## Commandline usage

### kvsd

```
kvsd 0.1.0
Benjamin Schilling <benjamin.schilling33@gmail.com>

USAGE:
    kvsd.exe [FLAGS] [OPTIONS]

FLAGS:
    -h, --help       Prints help information
        --tls        Set to enable TLS support for gRPC.
                     If set certificate and private key are expected as grpc.crt
                     and grpc.key in the execution directory of kvsc binary.
    -V, --version    Prints version information

OPTIONS:
        --ip <ip>        IP address the kvs daemon shall bind the gRPC interface to.
        --port <port>    Port the kvs daemon shall bind the gRPC interface to.
```

### kvsc

```
kvsc 0.1.0
Benjamin Schilling <benjamin.schilling33@gmail.com>

USAGE:
    kvsc.exe [FLAGS] [OPTIONS] [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
        --tls        Set to enable TLS support for gRPC.
                     If set certificate and private key are expected as ca.crt in the execution directory of kvsc
                     binary.
    -V, --version    Prints version information

OPTIONS:
        --ip <ip>        IP address the kvs daemon is bound to.
        --port <port>    Port the kvs daemon is bound to.

SUBCOMMANDS:
    delete    Delete the given key.
    get       Get the value of a given key.
    help      Prints this message or the help of the given subcommand(s)
    store     Store a given key value pair.
```

### Options

#### TLS

kvsd & kvsc supports TLS protected gRPC connections. 
To use this feature you have to run the binaries with the `--tls` option.
Before doing this you should have generated a valid key pair using the following command and placed the private key (`grpc.key`) as well as certificate (`grpc.crt`) in the same directory as the **ksvd** binary.
The **ksvc** binary requires the CA certificate used to sign the ksvd certficiate (`ca.crt`) in the same directory as the **ksvc** binary.
To  protect the web filesystem a reverse proxy like nginx should be used.

**Beware: The following command generates an insecure self-signed certificate and should be used for development only!**

> `openssl req -newkey rsa:2048 -nodes -keyout grpc.key -x509 -days 365 -out grpc.crt`

Afterwards the generated certificate (**in a non-development environment the CA certificate**) can be added as the CA certificate in **ksvc** or any gRPC client, like BloomRPC, to establish TLS protected connections.

## Building the project

### Development

#### Build

`cargo build`

The gRPC code is generated during build using the `build.rs` build script.

#### Run

The following example uses only the mandatory parts of the arguments, for a full example see below.

`cargo run --bin kvsd`

`cargo run --bin kvsc`

### Release

#### Build

`cargo build --release`

The gRPC code is generated during build using the `build.rs` build script.

## Architecture

> **Hint:**
> Use the **`Markdown Preview Enhanced`** vscode plugin to render the PlantUML drawings.

```puml
caption System Component Diagram

skinparam monochrome true
skinparam componentStyle uml2

scale max 650 width

component kvsd [
    <<component>>
    kvsd
]
component kvsc [
    <<component>>
    kvsc
]
component app [
    <<component>>
    Application
]

() ":27001\ngRPC" as gRPC


kvsd -- gRPC
kvsc -- gRPC
app -- gRPC

```

The kvsd is intended for embedded or industrial use cases where the hardware is limited but not too restricted. 
It can also be used in more powerful systems allowing more stored data.



## Backends

The kvsd supports two backends.

The first backend is implemented as a JSON array of key value pairs.
It is inteded for small databases of keys with small keys because the whole store is parsed and stored in RAM.

The second backend is implemented on file base.
Each entry is stored as a separate file with the file name representing the key and the content representing the value.

### JSON Backend

The JSON Backend stores all values in the following structure:

```json
{
    "key": "value",
    "key2": "value2",
    "key n": "value n"
}
```

On start-up of the kvsd the file is parsed and stored in a hashmap.
Therefore all data is stored in RAM.
Using this backend the length of the values is restricted to 1024 characters.
Otherwise the risk of consuming to much RAM during runtime is too high.
Additionally the number of key value pairs is restricted to 10.000 to prevent too much ressource consumption.
10.000 key value pairs result in around 11MB of RAM consumption.

#### Security

The JSON Backend is intended for less-secure environments.
The keys are stored in plain-text form, while only the values are encrypted.



The kvsd stores the data it receives in a JSON file.
This JSON file contains key value pairs.

```json
{
  "key": "derivation-value (32 byte)$iv (16 bytes)$encrypted-value",
  "key": "01234567890123450123456789012345$0123456789012345$0123456789012345$",
  "test": "testvalue"
}
```
The encrypted value is generated using the function `Base64(AES256GCM(Key, Initialization Vector, Plaintext))` as follows: 

> `Base64(AES256GCM(derive_key(dv), iv, plaintext value))`

```puml
```

### File Backend

The file backend stores each key value pair in a file.
Additionally a JSON file is created containing references to all files and the key related to each of them.

This allows to store dramatically bigger values.
Only the JSON file is loaded during runtime and the file is loaded, decrypted and returned only on request.

#### Security

The key is stored in an encrypted JSON file, containing a reference to an encrypted file (containing the value) and the secrect to decrypt the file.
The JSON file is decrypted with a secret derived from the 

```json
{
    "files": [
        {
            "key": "key",
            "filename": "file1",
            "secret": "key"
        },
        {
            "key": "key",
            "filename": "file2",
            "secret": "key"
        }
    ]
}
```

## License

SPDX-License-Identifier: MIT

## Copyright 

Copyright (C) 2020 Benjamin Schilling

debian/changelog
kvs (0.1.0-0) unstable; urgency=medium

   * Initial release 

 -- Benjamin Schilling <benjamin.schilling33@gmail.com>  Fri, 18 Dec 2020 15:55:00 +0100


src/kvsc/main.rs
/*
*  kvsc main
*  SPDX-License-Identifier: MIT
*  Copyright (C) 2020 Benjamin Schilling
*/

#![feature(proc_macro_hygiene, decl_macro)]

// Rust Standard Library
use std::io::{self, Read};

//tonic
use tonic::transport::{Certificate, ClientTlsConfig};

// gRPC imports
use kvs_api::kvs_client::KvsClient;
use kvs_api::KeyValuePair;
pub mod kvs_api {
    tonic::include_proto!("kvs_api");
}

//kvs crates
use utils::{crypto, filesystem_wrapper::get_exec_dir, input_validation};

// CLI interface
extern crate clap;
use clap::{App, AppSettings, Arg, SubCommand};

const INPUT_CLI: u8 = 0;
const INPUT_PIPE: u8 = 1;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Specify commandline arguments
    let matches = App::new("kvsc")
        .version(clap::crate_version!())
        .author("Benjamin Schilling <benjamin.schilling33@gmail.com>")
        .setting(AppSettings::SubcommandRequiredElseHelp)
        .arg(
            Arg::with_name("ip")
                .help("IP address the kvs daemon is bound to.")
                .required(false)
                .long("ip")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("port")
                .help("Port the kvs daemon is bound to.")
                .required(false)
                .long("port")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("tls")
                .help("Set to enable TLS support for gRPC.\nIf set certificate and private key are expected as ca.crt in the execution directory of kvsc binary.")
                .long("tls"),
        )
        .subcommand(
            SubCommand::with_name("store")
            .about("Store a given key value pair.")
            .arg(
                Arg::with_name("key")
                .long("key")
                .help("Key of the key value pair, max. length 32.")
                .takes_value(true)
                .required(true)
            )
            .arg(
                Arg::with_name("value")
                .long("value")
                .help("Value of the key value pair, max. length 1024.")
                .takes_value(true)
            )
            .arg(
                Arg::with_name("pipe")
                .long("pipe")
                .help("Allows piping the value in kvsc, alternative for \"value\".\n(No size restriction with file backend, \nbest used with base64 strings or complete files.)")
            )
        )
        .subcommand(
            SubCommand::with_name("get")
            .about("Get the value of a given key.")
            .arg(
                Arg::with_name("key")
                .long("key")
                .help("Key of the key value pair, max. length 32.")
                .takes_value(true)
                .required(true)
            )
        )
        .subcommand(
            SubCommand::with_name("delete")
            .about("Delete the given key.")
            .arg(
                Arg::with_name("key")
                .long("key")
                .help("Key of the key value pair, max. length 32.")
                .takes_value(true)
                .required(true)
            )
        )
        .get_matches();
    // Set IP and Port to default values
    let mut ip: String = "127.0.0.1".to_string();
    let mut port: String = "27001".to_string();
    // Set IP to provided value if existing
    if matches.is_present("ip") {
        if input_validation::validate_ipv4(matches.value_of("ip").unwrap().to_string()) {
            ip = matches.value_of("ip").unwrap().to_string();
        } else {
            eprintln!(
                "IP parameter {} invalid, only IPv4 allowed.",
                matches.value_of("ip").unwrap().to_string()
            );
            std::process::exit(0x0001);
        }
    }
    // Set Port to provided value if existing
    if matches.is_present("port") {
        if input_validation::validate_port(matches.value_of("port").unwrap().to_string()) {
            port = matches.value_of("port").unwrap().to_string();
        } else {
            eprintln!(
                "Port parameter {} invalid, only valid TCP port numbers allowed.",
                matches.value_of("port").unwrap().to_string()
            );
            std::process::exit(0x0001);
        }
    }

    // create a channel for the connection to the server
    let socket = format!("http://{}:{}", ip, port).parse().unwrap();
    let channel;
    if matches.is_present("tls") {
        let path = get_exec_dir();
        println!("TLS Option for gRPC given, looking for ca.crt in {}", path);
        let trust_store = match crypto::TrustStore::new(path) {
            Ok(trusted) => trusted,
            Err(e) => {
                eprintln!("Error during store: {:?}", e);
                std::process::exit(0x0001);
            }
        };
        let cert = Certificate::from_pem(trust_store.get_trusted_certificate());

        channel = tonic::transport::Channel::builder(socket)
            .tls_config(ClientTlsConfig::new().ca_certificate(cert))?
            .connect()
            .await?;
    } else {
        channel = tonic::transport::Channel::builder(socket).connect().await?;
    }

    // create a gRPC client from the channel
    let mut client = KvsClient::new(channel);

    // handle subcommands
    match matches.subcommand() {
        ("store", Some(sub_m)) => {
            // Perform input validation on options
            if !input_validation::validate_key(sub_m.value_of("key").unwrap().to_string()) {
                eprintln!("Provided key invalid.");
                std::process::exit(0x0001);
            }
            // Check whether either value or pipe are given
            let mut _value_input: u8 = INPUT_CLI;
            if sub_m.is_present("value") && !sub_m.is_present("pipe") {
                _value_input = INPUT_CLI;
            } else if !sub_m.is_present("value") && sub_m.is_present("pipe") {
                _value_input = INPUT_PIPE;
            } else {
                // If both value and pipe are given exit.
                eprintln!("Provide either \"value\" or \"pipe\".");
                std::process::exit(0x0001);
            }
            // for value perform input validation
            if _value_input == INPUT_CLI
                && !input_validation::validate_value(
                    sub_m.value_of("value").unwrap().to_string(),
                    false,
                )
            {
                eprintln!("Provided value invalid.");
                std::process::exit(0x0001);
            }

            // Get values of arguments
            let key = sub_m.value_of("key").unwrap().to_string();
            let mut value = String::new();
            if _value_input == INPUT_CLI {
                value = sub_m.value_of("value").unwrap().to_string();
            } else {
                match io::stdin().read_to_string(&mut value) {
                    Ok(string) => string,
                    Err(e) => {
                        eprintln!("Could not read value from stdin: {}.", e);
                        std::process::exit(0x0001);
                    }
                };
            }
            // creating a new Request
            let request = tonic::Request::new(KeyValuePair { key, value });
            // Send request and handle response
            match client.store(request).await {
                Ok(response) => {
                    println!("Storing key \"{}\" successful.", response.into_inner().key);
                    std::process::exit(0x0000);
                }
                Err(e) => {
                    eprintln!("Error during store: {:?}", e.message());
                    std::process::exit(0x0001);
                }
            };
        }
        ("get", Some(sub_m)) => {
            // Perform input validation on options
            if !input_validation::validate_key(sub_m.value_of("key").unwrap().to_string()) {
                eprintln!("Provided key invalid.");
                std::process::exit(0x0001);
            }
            // Get values of options
            let key = sub_m.value_of("key").unwrap().to_string();
            // creating a new Request
            let request = tonic::Request::new(KeyValuePair {
                key,
                value: "".to_string(),
            });
            // Send request and handle response
            match client.get(request).await {
                Ok(response) => {
                    println!("{}", response.into_inner().value);
                    std::process::exit(0x0000);
                }
                Err(e) => {
                    eprintln!("Error during get: {:?}", e.message());
                    std::process::exit(0x0001);
                }
            };
        }
        ("delete", Some(sub_m)) => {
            // Perform input validation on options
            if !input_validation::validate_key(sub_m.value_of("key").unwrap().to_string()) {
                eprintln!("Provided key or value invalid.");
                std::process::exit(0x0001);
            }
            // Get values of options
            let key = sub_m.value_of("key").unwrap().to_string();
            // creating a new Request
            let request = tonic::Request::new(KeyValuePair {
                key,
                value: "".to_string(),
            });

            // Send request and handle response
            match client.delete(request).await {
                Ok(response) => {
                    println!("Deleting key \"{}\" successful.", response.into_inner().key);
                    std::process::exit(0x0000);
                }
                Err(e) => {
                    eprintln!("Error during delete: {:?}", e.message());
                    std::process::exit(0x0001);
                }
            };
        }
        _ => {
            eprintln!("Unknown subcommand.");
            std::process::exit(0x0001);
        }
    };
}
